### PR TITLE
Initial version of the FrontierSilicon-Radio binding

### PR DESCRIPTION
**Pull request by Osray**
This binding can be used for controlling internet/mp3 radios based on a
frontier silicon radio chipset.
Tested with a Hama IR110 internet radio.

Disclaimer: I am not a Java developer ... :)

### CODE REVIEW

**Comment by teichsta** [comment]
Thanks @Osray for this contribution. FYI i'll hope the start the review in CW 5. Best, Thomas E.-E.

File: bundles/binding/org.openhab.binding.frontiersiliconradio/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+Bundle-Vendor: openHAB.org
+Bundle-Version: 1.7.0.qualifier
+Bundle-Activator: org.openhab.binding.frontiersiliconradio.internal.FrontierSiliconRadioActivator
+Bundle-ManifestVersion: 2
+Bundle-Description: This is the FrontierSiliconRadio binding of the open Home Aut
+ omation Bus (openHAB)
+Import-Package: org.apache.commons.httpclient;version="3.1.0",
+ org.apache.commons.httpclient.methods;version="3.1.0",
+ org.apache.commons.httpclient.params;version="3.1.0",
+ org.apache.commons.io;version="2.0.1",
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/META-INF/MANIFEST.MF]: please remove the version if you do not really rely on it

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+ * @since 1.7.0
+ */
+/**
+ * @author rainer
+ *
+ */
+/**
+ * @author rainer
+ *
+ */
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please fix this javadoc

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+	 * 
+	 * @return true when radio is on, false when radio is off
+	 */
+	public Boolean getPower() {
+		try {
+			FrontierSiliconRadioApiResult result = conn.doRequest("GET/netRemote.sys.power");
+			return result.getValueU8AsBoolean();
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+	 * 
+	 * @param powerOn true turns on the radio, false turns it off
+	 */
+	public void setPower(Boolean powerOn) {
+		try {
+			conn.doRequest("SET/netRemote.sys.power", "value="+(powerOn?"1":"0"));
+			return;
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+	 * @return volume: 0=muted, 32=max. volume
+	 */
+	public int getVolume() {
+		try {
+			FrontierSiliconRadioApiResult result = conn.doRequest("GET/netRemote.sys.audio.volume");
+			currentVolume = result.getValueU8AsInt();
+			return currentVolume;
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+	 * @param volume Radio volume: 0=mute, 32=max. volume
+	 */
+	public void setVolume( int volume ) {
+		try {
+			conn.doRequest("SET/netRemote.sys.audio.volume", "value="+volume );
+			currentVolume = volume;
+			return;
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+	 * 
+	 * @return operating mode. On hama radio: 0="Internet Radio", 1=Spotify, 2=Player, 3="AUX IN"
+	 */
+	public int getMode() {
+		try {
+			FrontierSiliconRadioApiResult result = conn.doRequest("GET/netRemote.sys.mode");
+			return result.getValueU32AsInt();
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+	 * 
+	 * @param mode On hama radio: 0="Internet Radio", 1=Spotify, 2=Player, 3="AUX IN"
+	 */
+	public void setMode( int mode ) {
+		try {
+			conn.doRequest("SET/netRemote.sys.mode", "value="+mode );
+			return;
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+	 * 
+	 * @return the station name, e.g. "WDR2"
+	 */
+	public String getPlayInfoName() {
+		try {
+			FrontierSiliconRadioApiResult result = conn.doRequest("GET/netRemote.play.info.name");
+			return result.getValueC8ArrayAsString();
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+	 * 
+	 * @return the radio info text, e.g. music title
+	 */
+	public String getPlayInfoText() {
+		try {
+			FrontierSiliconRadioApiResult result = conn.doRequest("GET/netRemote.play.info.text");
+			return result.getValueC8ArrayAsString();
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+	 */
+	public void setPreset(Integer presetId) {
+		try {		
+			conn.doRequest("SET/netRemote.nav.state", "value=1");
+			conn.doRequest("SET/netRemote.nav.action.selectPreset", "value="+presetId.toString() );
+			conn.doRequest("SET/netRemote.nav.state", "value=0");
+			return;
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+	 * 
+	 * @return true: radio is muted, false: radio is not muted
+	 */
+	public Boolean getMuted() {
+		try {
+			FrontierSiliconRadioApiResult result = conn.doRequest("GET/netRemote.sys.audio.mute");
+			return result.getValueU8AsBoolean();
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+	 * 
+	 * @param muted true: mute the radio, false: unmute the radio
+	 */
+	public void setMuted(Boolean muted) {
+		try {
+			conn.doRequest("SET/netRemote.sys.audio.mute", "value="+(muted?"1":"0"));
+			return;
+		}
+		catch (Exception e) {
+			logger.error("request failed");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadio.java]: please provide meaningful message (add url, more context)

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadioBinding.java [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+import org.openhab.core.types.State;
+import org.osgi.service.cm.ConfigurationException;
+import org.osgi.service.cm.ManagedService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+	
+
+/**
+ * Implement this class if you are going create an actively polling service
+ * like querying a Website/Device.
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadioBinding.java]: please remove this generic comment and add a meaningful description of the purpose of this class

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadioBinding.java [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+	public FrontierSiliconRadioBinding() {
+	}
+		
+	
+	public void activate() {
+	}
+	
+	public void deactivate() {
+		// deallocate resources here that are no longer needed and 
+		// should be reset when activating this binding again
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadioBinding.java]: remove this generic comment

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadioBinding.java [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+			}
+		}
+	}
+
+	/**
+	 * @{inheritDoc}
+	 */
+	@Override
+	protected void internalReceiveCommand(String itemName, Command command) {
+		logger.debug("internalReceiveCommand() is called!");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadioBinding.java]: remove this statement or decrease to TRACE level

File: bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadioBinding.java [hunk 17]
```diff
@@ -0,0 +1,382 @@
...
+		}	
+	}
+	
+	/**
+	 * @{inheritDoc}
+	 */
+	@Override
+	protected void internalReceiveUpdate(String itemName, State newState) {
+		
+	}
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.frontiersiliconradio/src/main/java/org/openhab/binding/frontiersiliconradio/internal/FrontierSiliconRadioBinding.java]: remove this overridden method

**Comment by teichsta** [comment]
Hi @Osray, sorry for the very late answer. Please find most of my review comments inline. In addition to that please take care of:

* please make sure each of you classes contains an appropriate description on class level and has all public methods documented
* update license header (using mvn license plugin)
* add your binding to distribution/pom.xml
* add your binding to bundles/binding/pom.xml

Thanks, Thomas E.-E.

**Comment by Osray** [comment]
Thomas, thanks for your comments. As i recently became father of twins, i haven't found the time to update the code since then. Sorry for the delay - hope this will change soon..

Rainer

**Comment by kaikreuzer** [comment]
Congrats, @Osray! You are now a really busy man, I guess :-)

**Comment by teichsta** [comment]
Congrats, @Osray! Just let me know, when you are through …

**Comment by paphko** [comment]
Nice! I almost started implementing a new binding for my Radio MD87180: http://internetradio.medion.com/
The API seems to be compatible to the Hama Radio :-)

So I merged the branch from Osray and I am currently revising the code. Expect a new pull request soon.

**Comment by teichsta** [comment]
sounds good … keep us posted about your progress!

**Comment by paphko** [comment]
FYI, I'm still struggeling with items updates without spamming the event bus once per second. Hopefully I'll find some time for further testing next weekend.

**Comment by teichsta** [comment]
superseded by #2576

**Comment by Osray** [comment]
Thanks @paphko for taking care of this binding and improving it. Much appreciated.

**Comment by paphko** [comment]
@Osray It's nice that I didn't have to start from scratch!
If you find some time, maybe you can also test the binding :-)

**Comment by flammy** [comment]
@Osray I have started an api-documentation here:

https://github.com/flammy/fsapi/blob/master/FSAPI.md

It would be nice if you share your knowledge about the fsapi.

**Pull request closed without merging**